packages/opentelemetry-plugin-mongodb-core/src/mongodb.ts

```diff
@@ -0,0 +1,254 @@
+ * limitations under the License.
+ */
+
+// mongodb.Server type is deprecated so every use trigger a lint error
+/* tslint:disable:deprecation */
+
+import { BasePlugin } from '@opentelemetry/core';
+import { Span, SpanKind, CanonicalCode } from '@opentelemetry/types';
+import {
+  Func,
+  MongoInternalCommand,
+  MongoInternalTopology,
+  AttributeNames,
+} from './types';
+import * as mongodb from 'mongodb';
+import * as shimmer from 'shimmer';
+
+/** MongoDB instrumentation plugin for OpenTelemetry */
+export class MongoDBPlugin extends BasePlugin<typeof mongodb> {
+  private readonly _SERVER_METHODS = ['insert', 'update', 'remove', 'command'];
+  private readonly _CURSOR_METHODS = ['_next', 'next'];
+
+  readonly supportedVersions = ['>=2 <3'];
+
+  constructor(readonly moduleName: string) {
+    super();
+  }
+
+  /**
+   * Patches MongoDB operations.
+   */
+  protected patch() {
+    this._logger.debug('Patched MongoDB');
+
+    if (this._moduleExports.Server) {
+      for (const fn of this._SERVER_METHODS) {
+        this._logger.debug(`patching mongodb-core.Server.prototype.${fn}`);
+        shimmer.wrap(
+          this._moduleExports.Server.prototype,
+          // Forced to ignore due to incomplete typings
+          // tslint:disable-next-line:ban-ts-ignore
+          // @ts-ignore
+          fn,
+          this._getPatchCommand(fn)
+        );
+      }
+    }
+
+    if (this._moduleExports.Cursor) {
+      this._logger.debug(
+        'patching mongodb-core.Cursor.prototype functions:',
+        this._CURSOR_METHODS
+      );
+      shimmer.massWrap(
+        [this._moduleExports.Cursor.prototype],
+        this._CURSOR_METHODS as never[],
+        // tslint:disable-next-line:no-any
+        this._getPatchCursor() as any
+      );
+    }
+
+    return this._moduleExports;
+  }
+
+  /** Unpatches all MongoDB patched functions. */
+  unpatch(): void {
+    shimmer.massUnwrap(
+      [this._moduleExports.Server.prototype],
+      this._SERVER_METHODS as never[]
+    );
+    shimmer.massUnwrap(
+      [this._moduleExports.Cursor.prototype],
+      this._CURSOR_METHODS as never[]
+    );
+  }
+
+  /** Creates spans for Command operations */
+  private _getPatchCommand(operationName: string) {
+    const plugin = this;
+    return (original: Func<mongodb.Server>) => {
+      return function patchedServerCommand(
+        this: mongodb.Server,
+        ns: string,
+        commands: MongoInternalCommand[] | MongoInternalCommand,
+        options: {} | Function,
+        callback: Func<unknown>
+      ): mongodb.Server {
+        const currentSpan = plugin._tracer.getCurrentSpan();
+        // @ts-ignore
+        const resultHandler =
+          typeof options === 'function' ? options : callback;
+        if (
+          currentSpan === null ||
+          typeof resultHandler !== 'function' ||
+          typeof commands !== 'object'
+        ) {
+          return original.apply(this, (arguments as unknown) as unknown[]);
```

Comment:
should we log here? This is the case where instrumentation is skipped because incorrect arguments are passed in correct?

Reply:
I don't think so since theorically it shouldn't happen, it's just safe check